### PR TITLE
[sping-boot-practice] Implemented add order endpoint  

### PR DESCRIPTION
**Pull request by Andriy-Hrynevetskyi**

### CODE REVIEW

File: src/main/java/bookstore/service/order/OrderServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+            orderItem.setBook(book);
+            orderItem.setOrder(order);
+            orderItem.setPrice(book.getPrice());
+            orderItemRepository.save(orderItem);
+            orderItems.add(orderItem);
+        }
+        order.setOrderItems(orderItems);
+        shoppingCartService.clearShoppingCart(shoppingCart);
+        shoppingCartRepository.save(shoppingCart);
+        return orderMapper.toDto(order);
```
Comment (TarasYurkovskyi) [inline src/main/java/bookstore/service/order/OrderServiceImpl.java]: i think u can split this method

File: src/main/resources/db/changelog/changes/12-create-orders-table.yaml [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        - createTable:
+            tableName: orders
+            columns:
+              - column:
+                  name: id
+                  type: bigint
+                  autoIncrement: true
+                  constraints:
+                    primaryKey: true
+                    nullable: false
```
Comment (TarasYurkovskyi) [inline src/main/resources/db/changelog/changes/12-create-orders-table.yaml:15]: u dont need nullable for primary key

**Review by TarasYurkovskyi — APPROVED** [review]
Good job!

File: src/main/java/bookstore/controller/OrderController.java
Comment (d1sam) [inline src/main/java/bookstore/controller/OrderController.java:1]: Available for users with role ADMIN
PATCH: /api/orders/{id}

Available for users with role USER
GET: /api/orders
POST: /api/orders
GET: /api/orders/{orderId}/items
GET: /api/orders/{orderId}/items/{itemId}

File: src/main/java/bookstore/controller/OrderController.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    @GetMapping("/{orderId}/items/{id}")
+    @Operation(summary = "Get certain order item")
+    public OrderItemDto getOrderItem(@PathVariable Long orderId,
+                                     @PathVariable Long id,
+                                     Authentication authentication) {
+        User user = (User)authentication.getPrincipal();
+        return orderItemService.getOrderItemByOrderId(user.getId(), orderId, id);
+    }
+
+    @PutMapping("/{id}")
```
Comment (d1sam) [inline src/main/java/bookstore/controller/OrderController.java]:
```suggestion
    @PatchMapping("/{id}")
```

File: src/main/java/bookstore/model/Order.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    private BigDecimal total;
+    @CreationTimestamp
+    @Column(nullable = false, name = "order_date")
+    private LocalDateTime orderDate;
+    @Column(nullable = false, name = "shipping_address")
+    private String shippingAddress;
+    @OneToMany(mappedBy = "order")
+    @ToString.Exclude
+    @EqualsAndHashCode.Exclude
+    private Set<OrderItem> orderItems;
```
Comment (d1sam) [inline src/main/java/bookstore/model/Order.java]:
```suggestion
    private Set<OrderItem> orderItems = new new HashSet<>();
```

File: src/main/java/bookstore/repository/orderitem/OrderItemRepository.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+public interface OrderItemRepository extends JpaRepository<OrderItem, Long> {
+    @Query("""
+        FROM OrderItem oi
+        JOIN FETCH oi.book
+        JOIN FETCH oi.order o
+        JOIN FETCH o.user u
+        WHERE o.id = :orderId
+        AND u.id = :userId
+            """)
+    List<OrderItem> findAllByOrderIdAndUserId(Long userId, Long orderId);
```
Comment (d1sam) [inline src/main/java/bookstore/repository/orderitem/OrderItemRepository.java]: what about `Pageable`?

File: src/main/java/bookstore/repository/orderitem/OrderItemRepository.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
+package bookstore.repository.orderitem;
+
+import bookstore.model.OrderItem;
+import java.util.List;
+import java.util.Optional;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+
+public interface OrderItemRepository extends JpaRepository<OrderItem, Long> {
```
Comment (d1sam) [inline src/main/java/bookstore/repository/orderitem/OrderItemRepository.java:13]:
```suggestion
@Repository
public interface OrderItemRepository extends JpaRepository<OrderItem, Long> {
```

File: src/main/java/bookstore/service/order/OrderServiceImpl.java (hunk 1 above)
Comment (d1sam) [reply src/main/java/bookstore/service/order/OrderServiceImpl.java]: agree, please divide logic in this method to smaller private methods.

File: src/main/java/bookstore/service/order/OrderServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+                        "Can't find shopping cart with given id: " + userId));
+        BigDecimal total = BigDecimal.ZERO;
+        for (CartItem cartItem : shoppingCart.getCartItems()) {
+            BigDecimal itemPrice = cartItem.getBook().getPrice()
+                    .multiply(BigDecimal.valueOf(cartItem.getQuantity()));
+            total = total.add(itemPrice);
+        }
+        Order order = new Order();
+        order.setUser(user);
+        order.setStatus(Order.Status.valueOf(Order.Status.PENDING.name()));
```
Comment (d1sam) [inline src/main/java/bookstore/service/order/OrderServiceImpl.java]:
```suggestion
        order.setStatus(Order.Status.PENDING);
```

**Review by d1sam — CHANGES_REQUESTED** [review]
good job, but some changes should be done!

**Review by vovasalyha — CHANGES_REQUESTED** [review]
Why do I see changes from other HW? Please fix it and request review one more time)

**Pull request merged**